### ISSUE TITLE
Warn if io.flutter.embedding.android.SplashScreenDrawable is used in android manifest. 

### ISSUE THREAD

**Issue by blaugold** [flutter/flutter#145663]
Give users a warning/error if io.flutter.embedding.android.SplashScreenDrawable is used since the support was removed. 

Original title: Elevated number of ANRs since upgrading to 3.16
Original Bug below: 

Since upgrading to Flutter 3.16.9 from 3.13.8 we are seeing a sharp increase in the number of ANRs reported by Google Play Console. Moving to Flutter 3.19 slightly decreased the number of ANRs. Rolling back to Flutter 3.13.8 is bringing the number of ANRs down quickly back to the previous rate. 

<img width="1648" alt="Screenshot 2024-03-24 at 21 57 59" src="https://github.com/flutter/flutter/assets/5929595/5cd5154e-74d8-40f0-8ad9-58d1e2ac6688">

Unfortunately, the increase is not contained in one ANR issue, but instead is spread across many issues with different stack traces of the blocked main thread. What is common to the ANR issues is that they all are reported with the designation “Input dispatching timed out (No focused window)”. At the same time, as the number of ANRs has increased, the number of slow cold start has as well. We interpret this as pointing to an issue early in the application lifecycle.

<img width="1648" alt="Screenshot 2024-03-24 at 21 58 15" src="https://github.com/flutter/flutter/assets/5929595/0fb12a3a-0034-4cd5-8944-3d4046131ec0">

Another commonality between the ANRs is that the main thread seems to be blocked in an JNI API call. Here are a few examples of the top most frames of the main thread:

```
  #00  pc 0x000000000005bebc  /apex/com.android.runtime/lib/bionic/libc.so (syscall+28)
  #01  pc 0x0000000000148f41  /apex/com.android.art/lib/libart.so (art::ConditionVariable::WaitHoldingLocks(art::Thread*)+88)
  #02  pc 0x00000000002ae1f3  /apex/com.android.art/lib/libart.so (art::JNI<false>::NewLocalRef(_JNIEnv*, _jobject*)+374)
  #03  pc 0x00000000001d21e9  /data/app/~~3TYf5A_pQiSEwNSXUPKQfQ==/de.lotum.whatsinthefoto.us-j4An6PoONWTcujGMYQ8bVg==/lib/arm/libflutter.so
  at android.os.MessageQueue.nativePollOnce (Native method)

  #00  pc 0x000000000004d5dc  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #01  pc 0x0000000000232cec  /apex/com.android.art/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks+140)
  #02  pc 0x0000000000557f34  /apex/com.android.art/lib64/libart.so (art::JNI<false>::NewLocalRef+1036)
  #03  pc 0x0000000003b9667c  /data/app/~~R7m6MW4DQL-PTFqli9wVZg==/com.google.android.trichromelibrary_631204033-vU_fZnhiQSVERDHwVaBsAA==/base.apk (BuildId: 686e9f12e405fac532a32ea4609fbedd2ac0cbbf)

  #00  pc 0x00000000000500dc  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #01  pc 0x0000000000232858  /apex/com.android.art/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks+140)
  #02  pc 0x00000000006b60c8  /apex/com.android.art/lib64/libart.so (art::JNI<false>::ReleaseStringChars+936)
  #03  pc 0x0000000000010c78  /apex/com.android.i18n/lib64/libicu_jni.so (PatternNative_compileImpl+344)
  at com.android.icu.util.regex.PatternNative.compileImpl (Native method)

  #00  pc 0x000000000005e378  /apex/com.android.runtime/lib/bionic/libc.so (syscall+28)
  #01  pc 0x0000000000114763  /apex/com.android.art/lib/libart.so (art::ConditionVariable::WaitHoldingLocks+86)
  #02  pc 0x00000000001ea391  /apex/com.android.art/lib/libart.so (artJniMethodEnd+224)
  #03  pc 0x00000000003323e9  /apex/com.android.art/lib/libart.so (art_jni_method_end+8)
  at android.content.res.AssetManager.nativeSetApkAssets (Native method)

  #00  pc 0x000000000007550c  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #01  pc 0x00000000001af800  /apex/com.android.art/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks(art::Thread*)+148)
  #02  pc 0x0000000000669f3c  /apex/com.android.art/lib64/libart.so (art::GoToRunnable(art::Thread*)+460)
  #03  pc 0x0000000000669d2c  /apex/com.android.art/lib64/libart.so (art::JniMethodEnd(unsigned int, art::Thread*)+28)
  at android.content.res.AssetManager.nativeOpenXmlAsset (Native method)

  #00  pc 0x000000000004db10  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+32)
  #01  pc 0x0000000000232cec  /apex/com.android.art/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks+140)
  #02  pc 0x0000000000568fc8  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallStaticIntMethodV+924)
  #03  pc 0x0000000002c91684  /data/app/~~R0xqUbLbkoyBG-rrbWtZhg==/com.google.android.trichromelibrary_626111933-roSPPDChy2ZOPOqKp69_bQ==/base.apk (BuildId: b6840e49b4ed341f5f079f8027835baddbb8ce2f)

  #00  pc 0x000000000004c4ac  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #01  pc 0x0000000000232858  /apex/com.android.art/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks+140)
  #02  pc 0x000000000045d3a0  /apex/com.android.art/lib64/libart.so (artQuickGenericJniEndTrampoline+1008)
  #03  pc 0x0000000000355640  /apex/com.android.art/lib64/libart.so (art_quick_generic_jni_trampoline+160)
  #04  pc 0x0000000002058f30  /memfd:jit-zygote-cache (android.os.BinderProxy.transact+1600)
```

It seems that lower end devices are affected at a much higher rate:

<img width="1648" alt="Screenshot 2024-03-24 at 21 54 02" src="https://github.com/flutter/flutter/assets/5929595/8e3b3462-994f-4910-9b31-8cede34e5f97">

Unfortunately, we have not found a way to reproduce these issues. We would appreciate any assistance in diagnosing this issue further!

**Comment by darshankawar** [comment]
@blaugold 
The logs seems to be coming from native rather than Flutter. 
Is there a way for you to symbolicate the crash using https://github.com/flutter/flutter/wiki/Crashes#get-the-symbols and provide them here so that we can take it a look further ?

**Comment by blaugold** [comment]
Sorry for the late response! 

The first example trace is the only one that contains a frame in the Flutter engine. Here is the symbolicated frame:

```
#03  pc 0x00000000001d21e9  /data/app/~~3TYf5A_pQiSEwNSXUPKQfQ==/de.lotum.whatsinthefoto.us-j4An6PoONWTcujGMYQ8bVg==/lib/arm/libflutter.so
/b/s/w/ir/cache/builder/src/out/android_release/../../third_party/android_tools/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/jni.h:560
```

**Comment by reidbaker** [comment]
Would you be willing to share the plugins you use? I am wondering if you have webview?

**Comment by blaugold** [comment]
@reidbaker Thanks for taking a look! Here is the list of our dependencies:

```
ab_testing_core: ^1.0.0
ab_testing_firebase: ^1.0.0
args: ^2.4.2
async: ^2.11.0
cbl_flutter_ee: 3.0.0-dev.2
cbl_flutter: 3.0.0-dev.2
cbl_sentry: 2.0.0-dev.3
cbl: 3.0.0-dev.2
collection: ^1.17.0
connectivity_plus: ^5.0.2
crypto: ^3.0.3
csv: ^6.0.0
fading_edge_scrollview: ^4.1.1
firebase_analytics: ^10.8.0
firebase_auth: ^4.16.0
firebase_core: ^2.24.0
firebase_crashlytics: ^3.4.0
firebase_dynamic_links: ^5.4.0
firebase_messaging: ^14.7.0
firebase_performance: ^0.9.3
firebase_remote_config: ^4.3.0
flutter_countdown_timer: ^4.1.0
flutter_facebook_auth: 4.4.1+1
flutter_local_notifications:
flutter_mobx: ^2.2.1+1
flutter_spinkit: ^5.2.1
gap: ^3.0.1
google_mobile_ads: 4.0.0
google_sign_in: ^6.1.6
hive: ^2.2.3
html: ^0.15.4
http: ^1.1.0
in_app_purchase_android: ^0.3.2
in_app_purchase_storekit: ^0.3.8+1
in_app_purchase: ^3.1.13
intl: ^0.19.0
js: ^0.6.7
lottie_native: ^0.1.2
meta: ^1.9.1
mobx: ^2.3.0+1
package_info_plus: ^4.2.0
path_provider: ^2.1.2
path: ^1.8.3
provider: ^6.1.2
rate_my_app: ^2.0.0
rive: ^0.13.1
sentry_flutter: ^7.18.0
sentry: ^7.18.0
sign_in_with_apple: ^5.0.0
soundpool:
sqflite: ^2.3.2
super_tooltip: ^2.0.7
survicate_flutter_sdk:
tar: ^1.0.3
timezone: ^0.9.2
universal_html: ^2.2.4
url_launcher: ^6.2.4
usercentrics_sdk: 2.13.2
uuid: ^3.0.7
vector_math: ^2.1.4
webview_flutter: ^4.4.2
yaml: ^3.1.2
```

We are using web views directly in a few places, but we also show ads, and I suspect the web views used for those far outweigh our direct use.

Until recently, we also used `isolate_handler` for accessing plugins from background isolates. This package creates a new Flutter engine to work around the old limitations for background isolates regarding plugins. Now that background isolates in the same engine can access plugins, we have migrated away from `isolate_handler`. I have a hunch that multiple Flutter engines could cause a deadlock, but we haven't yet tried upgrading to 3.16 again after the migration.

**Comment by reidbaker** [comment]
@zanderso do you know where we have documentation around the behavior of multiple engines and what should or should not be expected?

**Comment by zanderso** [comment]
@gaaclarke and/or @xster may be able to give advice about multiple engines.

**Comment by gaaclarke** [comment]
I think the OP should use Android Studio to profile their app and look for any blocks on the platform thread (UI thread in Android parlance) that are long.  The fact that it shows up more in old devices may mean something slow is happening and it's not a deadlock.

> do you know where we have documentation around the behavior of multiple engines and what should or should not be expected?

Do you mean like "Don't do X or you'll get a deadlock."?  There is no guidance since they were designed to be used with the same semantics as a regular engine.   Obviously spawning 10s of engines in one event may tip you over the limit, just as any expensive operation would.

**Comment by carl-marc** [comment]
@blaugold did you get anywhere with this?
I have the same problem too.

**Comment by jonahwilliams** [comment]
The stack traces we have on this issue don't indicate a problem in the flutter engine itself, and without any ability to reproduce the problem its unlikely we'll make any progress. I'm going to close the issue, but if you find more information feel free to reopen or file a new bug.

**Comment by blaugold** [comment]
I think we have identified the issue. 

Between Flutter 3.13 and 3.16 support for `io.flutter.embedding.android.SplashScreenDrawable` has been removed. We specified the same drawable through `SplashScreenDrawable` and as the background of `LaunchTheme`. So when support for `io.flutter.embedding.android.SplashScreenDrawable` was removed, nothing visibly changed. 

We also do quite a bit of initialization in `main` before calling `runApp`. During this time, the app produces no frames, which can be interpreted by Android as an ANR on slow devices, for example. `io.flutter.embedding.android.SplashScreenDrawable` masked this by drawing the splash screen drawable between the app launching and our Flutter app producing the first frame.

Our likely solution is to redraw the splash screen in Flutter at the start of `main`. I'll report back once we have rolled out this change.
